# Student Performance Analysis

## Objective

This notebook aims to analyze student behavior in learning management systems as a predictor of learning success. The study will employ data mining techniques with different feature sets, including LMS data, course characteristics, and student activities, to predict student performance.

### Goals

1. Determine which data sources yield the highest accuracy in predicting student performance.
2. Investigate whether students with high activity levels outperform those with lower activity levels.

## Import Libraries

In [57]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set(color_codes = True)
import warnings

warnings.filterwarnings('ignore')
%matplotlib inline

## Load Dataset

**Let us read the data from an excel file**, <br>
We will print top few rows to understand about the various data columns

In [58]:
# Load the Excel file into a DataFrame
file_path = '2022-23 CST4070 Applied Data Analytics - Tools, Practical Big Data Handling, Cloud Distribution Grad.xlsx'
df = pd.read_excel(file_path)

df.head()

UserID  Campus Quiz: Formative test w1 (Real)  \
0   user33   Dubai                              -   
1  user172   Dubai                              -   
2   user57  Hendon                              -   
3   user93   Dubai                              -   
4  user158   Dubai                              -   

  Quiz: Formative test w1 (Jan) (Real) Quiz: Formative test w2 (Real)  \
0                                    -                              -   
1                                    -                              -   
2                                    -                              -   
3                                    -                              -   
4                                    -                              -   

  Quiz: Formative test w2 Jan31 (Real) Quiz: Formative test w3 (Real)  \
0                                    -                              -   
1                                    -                              -   
2                                    9                              -   
3                                    -                              -   
4                                    -                              -   

  Quiz: Formative test w3 Jan 31 (Real) Quiz: Formative test w4 (Real)  \
0                                     -                              -   
1                                     -                              -   
2                                     6                              -   
3                                     -                              -   
4                                     -                              -   

  Quiz: Formative test w4 Feb. 14 (Real)  ...  \
0                                      -  ...   
1                                      -  ...   
2                                   5.33  ...   
3                                      -  ...   
4                                      -  ...   

  Assignment: Component 6 - Individual Submission link (Real).1  \
0                                                  -              
1                                                  -              
2                                                  -              
3                                                  -              
4                                                  -              

  Quiz: Formative test week 9 (Real)  \
0                                  -   
1                                  -   
2                                  -   
3                                  -   
4                                  -   

  Quiz: Formative test week 9 (March 21) (Real)  \
0                                             -   
1                                             -   
2                                             -   
3                                             -   
4                                             -   

  Assignment: Appendix for Dubai CW1 (Real) Quiz: Formative test w10 (Real)  \
0                                         -                               -   
1                                         -                               -   
2                                         -                               -   
3                                         -                               -   
4                                         -                               -   

  Assignment: Submit your predictions here (Real)  \
0                                               -   
1                                               -   
2                                               -   
3                                               -   
4                                               -   

  Assignment: Submit your notebook here (Real)  \
0                                            -   
1                                            -   
2                                            -   
3                                            -   
4                                            -   

  Assignment: Compone

In [59]:
df.shape

(174, 75)

In [60]:
df.UserID.nunique()

174

The dataset contains multiple columns related to quizzes, assignments, and other components of the course with a total of 174 rows. Each row represents a student's performance in these activities.

In [61]:
# Filtering records where the campus is 'Hendon'
df_hendon = df[df['Campus'] == "Hendon"]

# Displaying the first few rows of the filtered DataFrame
df_hendon.head()

UserID  Campus Quiz: Formative test w1 (Real)  \
2    user57  Hendon                              -   
6    user61  Hendon                              -   
8   user111  Hendon                              -   
10   user97  Hendon                           3.33   
12  user120  Hendon                           6.67   

   Quiz: Formative test w1 (Jan) (Real) Quiz: Formative test w2 (Real)  \
2                                     -                              -   
6                                     -                              -   
8                                     -                              -   
10                                    -                              5   
12                                    -                           4.33   

   Quiz: Formative test w2 Jan31 (Real) Quiz: Formative test w3 (Real)  \
2                                     9                              -   
6                                     2                              -   
8                                     6                              -   
10                                    -                              7   
12                                    -                              3   

   Quiz: Formative test w3 Jan 31 (Real) Quiz: Formative test w4 (Real)  \
2                                      6                              -   
6                                      9                              -   
8                                      4                              -   
10                                     -                           3.33   
12                                     -                              -   

   Quiz: Formative test w4 Feb. 14 (Real)  ...  \
2                                    5.33  ...   
6                                    4.33  ...   
8                                       1  ...   
10                                      -  ...   
12                                      -  ...   

   Assignment: Component 6 - Individual Submission link (Real).1  \
2                                                   -              
6                                                   -              
8                                                   -              
10                                                  -              
12                                                  -              

   Quiz: Formative test week 9 (Real)  \
2                                   -   
6                                   -   
8                                   -   
10                                  -   
12                                  -   

   Quiz: Formative test week 9 (March 21) (Real)  \
2                                              -   
6                                              8   
8                                              -   
10                                             -   
12                                             -   

   Assignment: Appendix for Dubai CW1 (Real) Quiz: Formative test w10 (Real)  \
2                                          -                               -   
6                                          -                               -   
8                                          -                               -   
10                                         -                            2.67   
12                                         -                               -   

   Assignment: Submit your predictions here (Real)  \
2                                                -   
6                                                -   
8                                                -   
10                                               -   
12                                               -   

   Assignment: Submit your notebook here (Real)  \
2                                             -   
6                                             -   
8                                             -   
10                                            -   
12      

In [62]:
df_hendon.shape

(116, 75)

The filtered dataset contains multiple columns related to quizzes, assignments, and other components of the course, but only for students at the Hendon campus with a total of 116 rows

In [63]:
# File path for the second dataset
file_path_logs = 'logs.csv'

# Loading the second dataset into a DataFrame
df_logs = pd.read_csv(file_path_logs)

# Displaying the first few rows of the DataFrame
df_logs.head()

Time  UserID                                      Event context  \
0  18/06/23, 13:41  user25  Course: 2022-23 CST4070 Applied Data Analytics...   
1  18/06/23, 13:38  user25  Course: 2022-23 CST4070 Applied Data Analytics...   
2  18/06/23, 02:02  user58  Course: 2022-23 CST4070 Applied Data Analytics...   
3  17/06/23, 20:17  user18  Course: 2022-23 CST4070 Applied Data Analytics...   
4  17/06/23, 20:07  user18  Course: 2022-23 CST4070 Applied Data Analytics...   

  Component     Event name                                        Description  
0    System  Course viewed  The user with id '138154' viewed the course wi...  
1    System  Course viewed  The user with id '138154' viewed the course wi...  
2    System  Course viewed  The user with id '156217' viewed the course wi...  
3    System  Course viewed  The user with id '141836' viewed the course wi...  
4    System  Course viewed  The user with id '141836' viewed the course wi...

In [64]:
df_logs.shape

(113725, 6)

The logs dataset contains information about user activities, including the time of the activity, the user ID, the event context, the component involved, the event name, and a description of the event. The loag dataset contains a record of 113,725 records

In [65]:
# Joining the two DataFrames using the 'UserID' column
df_joined = pd.merge(left=df_hendon, right=df_logs, left_on='UserID', right_on='UserID')

# Displaying the first few rows of the joined DataFrame
df_joined.head()

UserID  Campus Quiz: Formative test w1 (Real)  \
0  user57  Hendon                              -   
1  user57  Hendon                              -   
2  user57  Hendon                              -   
3  user57  Hendon                              -   
4  user57  Hendon                              -   

  Quiz: Formative test w1 (Jan) (Real) Quiz: Formative test w2 (Real)  \
0                                    -                              -   
1                                    -                              -   
2                                    -                              -   
3                                    -                              -   
4                                    -                              -   

  Quiz: Formative test w2 Jan31 (Real) Quiz: Formative test w3 (Real)  \
0                                    9                              -   
1                                    9                              -   
2                                    9                              -   
3                                    9                              -   
4                                    9                              -   

  Quiz: Formative test w3 Jan 31 (Real) Quiz: Formative test w4 (Real)  \
0                                     6                              -   
1                                     6                              -   
2                                     6                              -   
3                                     6                              -   
4                                     6                              -   

  Quiz: Formative test w4 Feb. 14 (Real)  ...  \
0                                   5.33  ...   
1                                   5.33  ...   
2                                   5.33  ...   
3                                   5.33  ...   
4                                   5.33  ...   

  Assignment: Submit your predictions here (Real)  \
0                                               -   
1                                               -   
2                                               -   
3                                               -   
4                                               -   

  Assignment: Submit your notebook here (Real)  \
0                                            -   
1                                            -   
2                                            -   
3                                            -   
4                                            -   

  Assignment: Component 6 - Group PDF Submission link (Real)  \
0                                                  -           
1                                                  -           
2                                                  -           
3                                                  -           
4                                                  -           

  Assignment: Component 6 - Group Screenshot Submission link (Real)  \
0                                                  -                  
1                                                  -                  
2                                                  -                  
3                                                  -                  
4                                                  -                  

  Course total (Real)             Time  \
0               61.13  13/06/23, 16:08   
1               61.13  13/06/23, 16:08   
2               61.13  13/06/23, 16:07   
3               61.13  13/06/23, 15:23   
4               61.13  13/06/23, 15:22   

                                       Event context Component  \
0  Course: 2022-23 CST4070 Applied Data Analytics...    System   
1                             Folder: Lecture slides    Folder   
2  Course: 2022-23 CST4070 Applied Data Analytics...    System   
3                             Folder: Lecture slides    Folder   
4  Course: 2022-23 CST4070 Applied Data Analytics...    S

In [66]:
df_joined.shape

(100058, 80)

In [67]:
df_joined.UserID.nunique()

116

The joined DataFrame contains columns from both the original DataFrames, allowing us to analyze student performance in relation to their activity logs. It contains 100,058 records

## Data Cleaning

Before diving into the analysis, it's crucial to clean the data. This involves:

1. Removing duplicates
2. Handling missing values
3. Converting data types, if necessary
4. Renaming columns for better readability

Let's start by examining the missing values in both datasets.

### Removing Duplicates


In [68]:
# Checking for duplicate records in the DataFrame
duplicates = df_joined.duplicated().sum()
duplicates

16149

It appears that there are 16,149 duplicate records in the joined DataFrame. These duplicates need to be removed to ensure the quality of the data.

In [69]:
# Removing duplicate records from the DataFrame
df_joined_clean = df_joined.drop_duplicates()

# Verifying that duplicates have been removed
remaining_duplicates = df_joined_clean.duplicated().sum()
remaining_duplicates

0

All 16,149 duplicate records have been successfully removed, leaving us with a clean dataset

### Handling Missing Values

In [71]:
# Checking for missing values in the DataFrame
missing_values = df_joined.isnull().sum()
missing_values[missing_values > 0]

Series([], dtype: int64)

It appears that there are no missing values in the joined DataFrame.

### Data Type Conversion

In [72]:
# Checking the data types of each column in the DataFrame
data_types = df_joined_clean.dtypes
data_types

UserID                                  object
Campus                                  object
Quiz: Formative test w1 (Real)          object
Quiz: Formative test w1 (Jan) (Real)    object
Quiz: Formative test w2 (Real)          object
                                         ...  
Time                                    object
Event context                           object
Component                               object
Event name                              object
Description                             object
Length: 80, dtype: object